### PR TITLE
Fix double free in loader destructor and race condition in consumer/producer

### PR DESCRIPTION
**Pull request by jantonguirao**
The problem was a race condition with using atomic counter by several worker threads. Using fetch_add to resolve the race condition. 
Bonus: Using smart pointers to prevent any more double deletes to ever happen.

Signed-off-by: Joaquin Anton <janton@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/reader/loader/loader.h [hunk 1]
```diff
@@ -122,23 +122,22 @@ class Loader {
       for (int i = 0; i < initial_empty_size_; ++i) {
         LoadTarget* tensor = new LoadTarget();
```
Comment (JanuszL) [inline dali/pipeline/operators/reader/loader/loader.h]: Why we are not using any smart pointers for that? In that case we would avoid calling delete directly on them.

Comment (jantonguirao) [reply dali/pipeline/operators/reader/loader/loader.h]: That's exactly what I said this morning :) But we are trying to fix the underlying problem and adding the shared pointer for testing now will just hide the problem.
Let's change it once we figure out the issue

Comment (JanuszL) [reply dali/pipeline/operators/reader/loader/loader.h]: Understood

File: dali/pipeline/operators/reader/loader/loader.h [hunk 2]
```diff
@@ -107,6 +106,7 @@ class Loader {
     // perform an iniital buffer fill if it hasn't already happened
     if (!initial_buffer_filled_) {
       TimeRange tr("[Loader] Filling initial buffer", TimeRange::kBlue1);
+      std::lock_guard<std::mutex> lock(return_mutex_);
```
Comment (awolant) [inline dali/pipeline/operators/reader/loader/loader.h]: Maybe we should add a comment that this is about `empty_tensors_` couple lines later? So we remember later

File: dali/pipeline/operators/reader/loader/loader.h [hunk 3]
```diff
@@ -185,7 +188,7 @@ class Loader {
   }
   std::vector<LoadTarget*> sample_buffer_;
```
Comment (jantonguirao) [inline dali/pipeline/operators/reader/loader/loader.h]: should probably be `std::vector<std::shared_ptr<LoadTarget>>`

File: dali/pipeline/operators/reader/loader/loader.h (hunk 2 above)
Comment (jantonguirao) [reply dali/pipeline/operators/reader/loader/loader.h]: what comment do you have in mind?

File: dali/pipeline/operators/reader/loader/loader.h (hunk 1 above)
Comment (jantonguirao) [reply dali/pipeline/operators/reader/loader/loader.h]: done

File: dali/pipeline/operators/reader/loader/loader.h (hunk 3 above)
Comment (jantonguirao) [reply dali/pipeline/operators/reader/loader/loader.h]: done

**Comment by jantonguirao** [comment]
@Kh4L fixed race condition

**Comment by jantonguirao** [comment]
Build 679910

**Comment by jantonguirao** [comment]
Build 679960

File: dali/pipeline/operators/reader/file_reader_op.h [hunk 4]
```diff
@@ -30,24 +30,24 @@ class FileReader : public DataReader<CPUBackend, ImageLabelWrapper> {
   void RunImpl(SampleWorkspace *ws, const int i) override {
     const int idx = ws->data_idx();
 
-    auto* image_label = GetSample(idx);
+    auto& image_label = GetSample(idx);
```
Comment (Kh4L) [inline dali/pipeline/operators/reader/file_reader_op.h]:
```suggestion
    const auto& image_label = GetSample(idx);
```

File: dali/pipeline/operators/reader/caffe2_reader_op.h [hunk 5]
```diff
@@ -30,13 +30,7 @@ class Caffe2Reader : public DataReader<CPUBackend, Tensor<CPUBackend>> {
...
 
   void RunImpl(SampleWorkspace* ws, const int i) override {
-    const int idx = ws->data_idx();
-
-    auto* raw_data = GetSample(idx);
-
-    parser_->Parse(*raw_data, ws);
-
-    return;
+    parser_->Parse(GetSample(ws->data_idx()), ws);
```
Comment (mzient) [inline dali/pipeline/operators/reader/caffe2_reader_op.h]: The code does the same in 7x fewer lines, but keeping these intermediate variables might facilitate debugging.

File: dali/pipeline/operators/reader/file_reader_op.h (hunk 4 above)
Comment (jantonguirao) [reply dali/pipeline/operators/reader/file_reader_op.h]: done

Comment (jantonguirao) [reply dali/pipeline/operators/reader/file_reader_op.h]: done

**Pull request merged**